auth-gateway/index.js

```diff
@@ -0,0 +1,162 @@
+require('dotenv').config();
+
+const polka = require('polka');
+const jwksRsa = require('jwks-rsa');
+const jwt = require('jsonwebtoken');
+
+const { AUTH0_DOMAIN, AUTH0_AUDIENCE } = process.env;
+const AUTH0_ISSUER = `https://${AUTH0_DOMAIN}/`;
+
+// Grabs public key, caches it, from Auth0's servers
+// asymmetric signature used to ensure that not only the jwt was unmodified
+// since signed, but that the issuer was Auth0 (since they hold the private key)
+const jwksFn = jwksRsa({
+  cache: true,
+  rateLimit: true,
+  jwksRequestsPerMinute: 5,
+  jwksUri: `https://${AUTH0_DOMAIN}/.well-known/jwks.json`
+});
+
+// Ref: https://github.com/auth0/node-jwks-rsa/blob/master/src/integrations/express.js
+function getSecretKey(header, cb) {
+  jwksFn.getSigningKey(header.kid, (err, key) => {
+    if (err) {
+      cb(err, null);
+      return;
+    }
+
+    cb(null, key.publicKey || key.rsaPublicKey);
+  });
+}
+
+// Verifies signature, as well as the issuer and audience, to guarantee
+// that the token was truly for our resource server
+const verifyToken = token =>
+  new Promise((resolve, reject) => {
+    jwt.verify(
+      token,
+      getSecretKey,
+      {
+        issuer: AUTH0_ISSUER,
+        audience: AUTH0_AUDIENCE,
+        clockTolerance: 2, // seconds to deal with clock skew
+        algorithms: ['RS256']
+      },
+      (err, dToken) => {
+        if (err) {
+          console.error(err);
+          return reject(err);
+        }
+
+        return resolve(dToken);
+      }
+    );
+  });
+
+const bearerPrefix = 'Bearer ';
+const bearerPrefixLen = bearerPrefix.length;
+
+const cookiePrefix = 'access_token=';
+const cookieOffset = cookiePrefix.length;
+
+const getAuthToken = req => {
+  let whereFoundCount = 0;
+
+  let bearerIdx;
+  let cookieIdx;
+
+  if (req.headers.cookie) {
+    cookieIdx = req.headers.cookie.indexOf(cookiePrefix);
+
+    if (cookieIdx > -1) {
+      whereFoundCount += 1;
+    }
+  }
+
+  if (req.query.access_token) {
+    whereFoundCount += 1;
+  }
+
+  if (req.headers.authorization) {
+    bearerIdx = req.headers.authorization.indexOf(bearerPrefix);
+
+    if (bearerIdx > -1) {
+      whereFoundCount += 1;
+    }
+  }
+
+  // oauth2 spec specifies token should be presented in only one location
```

Comment:
Could you point me to the section for this? It doesn't seem like we "MUST" return an error / forbidden. I guess I'm OK checking anyway though, since it seems like it will find client errors faster.

Reply:
I'll look for the reference